doc/config_yang_validation/config_db_yang_validation.md

```diff
@@ -0,0 +1,178 @@
+![](../../images/config_yang_validation/yang_portchannel.png)
+
+For PortChannel and many of the config fields listed above, the field validation is unnecessarily duplicated in sonic-utilities/config and in YANG models. The goal of this project is to utilize preexisting YANG models to validate config CLI field updates, without unnecessarily defining separate, redundant validation specs ad-hoc in sonic-utilities. 
+
+During the migration process, we will first leave the ad-hoc validation code in-place, and leave an option to configure type of validation used: ad-hoc or YANG validation. Once YANG validation has stabilized and YANG validation coverage has widened, we will remove the preexisting ad-hoc validation code for non performance-sensitive scenarios. 
+
+In performance-sensitive scenarios, such as warm and fast reboot, ad-hoc validation will remain the mode of validation, as GCU is more time-intensive.
+
+
+## Design
+
+### Design Summary
+Currently, updates to ConfigDB are published via 2 steps:
+1.	ad-hoc ConfigDB field validation 
+2.	Call to config_db.set_entry(target_value) which blindly pushes the target_value passed via connection to ConfigDB using ConfigDBConnector object
+
+Where YANG models are applicable, this project replaces these two steps with a ConfigDBConnector class decorator which validates the new config using preexisting YANG models before pushing the new config update to ConfigDB.
+
+Why use ConfigDBConnector class decorator:
+- 	allows us to preserve the preexisting instance of ConfigDBConnector (no new object initialization is required)
+- 	decorates only the necessary instances of ConfigDBConnector with added YANG validation
+- 	Non-invasive, all ConfigDBConnector code not writing to ConfigDB (no YANG validation necessary) can be left as-is
+
+### Decorator Graphical Depiction
+Existing ConfigDBConnector class and methods-
+
+![](../../images/config_yang_validation/decorator1.png)
+
+Decorator only overrides contents of original set_entry() method to include YANG validation
+
+![](../../images/config_yang_validation/decorator2.png)
+
+### YANG Validation Implementation
+The YANG validation included in our ConfigDBConnector set_entry() decorator leverages the preexisting SONiC Generic Config Updater library, which uses YANG to validate JSON patches that contain write requests to ConfigDB, and then pushes these updates to ConfigDB only after YANG validation passes. 
+
+Sample JSON patch:
+
+![](../../images/config_yang_validation/jsonpatch.png)
+
+### Decorator Pseudo Code
+```
+def validate_decorator(config_db_connector):
+        def validated_set_entry(updated_value):
+                update_json_patch = updated_value.to_json_patch()
+                GenericConfigUpdater.apply_patch(update_json_patch)
+        config_db_connector.set_entry = validated_set_entry
+
+```
+For the buffermgrd C++ use case, there are 3 options for YANG validation implementation:
```

Comment:
Can you highlight which among the 3 options will be considered?

Reply:
Yes, I added a note about our most likely method of implementation